File: jobs/openalex_parser.py

from pymongo import MongoClient
import configparser
import os

from diophila import OpenAlex
from nameparser import HumanName
from Levenshtein import ratio


class OpenAlexParser():
    def __init__(self) -> None:
        self.TYPES = {
            "book-section": "chapter",
            "monograph": "book",
            "report-component": "others",
            "report": "others",
            "peer-review": "others",
            "book-track": "book",
            "journal-article": "article",
            "book-part": "book",
            "other": "others",
            "book": "book",
            "journal-volume": "article",
            "book-set": "book",
            "reference-entry": "others",
            "proceedings-article": "others",
            "journal": "others",
            "component": "others",
            "book-chapter": "chapter",
            "proceedings-series": "others",
            "report-series": "others",
            "proceedings": "others",
            "database": "others",
            "standard": "others",
            "reference-book": "book",
            "posted-content": "others",
            "journal-issue": "others",
            "dissertation": "dissertation",
            "grant": "others",
            "dataset": "others",
            "book-series": "book",
            "edited-book": "book",
        }


        # read the config file
        config = configparser.ConfigParser()
        path = os.path.dirname(__file__)
        config.read(os.path.join(path, 'config.ini'))

        self.inst_id = config['OpenAlex']['Institution']
        self.startyear = config['DEFAULT']['StartYear']

        # set up database connection
        client = MongoClient(config['Database']['Connection'])
        self.osiris = client[config['Database']['Database']]

        # osiris['queue'].delete_many({})


        # set up OpenAlex
        self.openalex = OpenAlex(config['DEFAULT'].get('AdminMail'))


    def getUserId(self, name, orcid=None):
        if orcid:
            user = self.osiris['users'].find_one({'orcid': orcid})
            if user:
                return user['_id']
        user = self.osiris['users'].find_one(
            {'last': name.last, 'first': {'$regex': '^'+name.first+'.*'}})
        if user:
            return user['_id']
        return None

    def getAbstract(self, inverted_abstract):
        if not inverted_abstract: return None
        
        abstract = []
        for word in inverted_abstract:
            occurence = inverted_abstract[word]
            for oc in occurence:
                abstract.append((oc, word))
        abstract = " ".join([i[1] for i in sorted(abstract)])
        return abstract

    def getJournal(self, issn):
        journal = self.osiris['journals'].find_one({'issn': {'$in': issn}})
        if journal:
            return journal

        # if journal does not exist: create one
        source = self.openalex.get_single_venue(issn[-1], "issn")
        if not source or source['type'] != 'journal':
            return None

        new_journal = {
            'journal': source['display_name'],
            'abbr': source['abbreviated_title'],
            'publisher': source['host_organization_name'],
            'issn': source['issn'],
            'oa': source['is_oa'],
            'openalex': source['id'].replace('https://openalex.org/', '')
        }
        new_doc = self.osiris['journals'].insert_one(new_journal)

        new_journal['_id'] = new_doc.inserted_id
        return new_journal

    def get_works(self):
        # NOPE: use created_date and updated_date to filter
        # Not possible, needs payed version

        filters = {
            "from_publication_date": self.startyear + "-01-01",
            "institutions.id": self.inst_id,
            "has_doi": 'true',
            # 'doi': '10.12688/f1000research.111175.1'
        }

        
        possible_dupl = self.osiris['activities'].find({
            'type': 'publication',
                    'year': {'$gte': int(self.startyear)},
        }, {'title': 1})
        possible_dupl = [
            (i['_id'], i['title']) for i in possible_dupl
        ]

        pages_of_works = self.openalex.get_list_of_works(filters=filters, pages=None)

        for page in pages_of_works:
            for work in page['results']:
                if work['is_retracted']:
                    continue

                # print(work['doi'])
                if not work['doi'] or 'https://doi.org/' not in work['doi']:
                    continue

                pubmed = work['ids'].get('pmid')
                if pubmed:
                    pubmed = pubmed.replace('https://pubmed.ncbi.nlm.nih.gov/', '')

                # check if element is in the database
                doi = work['doi'].replace('https://doi.org/', '')
                if self.osiris["activities"].count_documents({
                    '$or': [
                        {'doi': doi},
                        {'pubmed': pubmed}
                    ]
                }) > 0:
                    continue
                if self.osiris['queue'].count_documents({'doi': doi}) > 0:
                    continue
                print(doi)
                # print(work['title'])

                typ = self.TYPES.get(work['type'])

                authors = []
                for a in work['authorships']:
                    # match via name and ORCID
                    name = HumanName(a['author']['display_name'])
                    orcid = a['author'].get('orcid')
                    if (orcid):
                        orcid = orcid.replace('https://orcid.org/', '')

                    user = self.getUserId(name, orcid)
                    pos = a['author_position']
                    if pos == 'middle' and a.get('is_corresponding'):
                        pos = 'corresponding'

                    authors.append({
                        'last': name.last,
                        'first': name.first + (' ' + name.middle if name.middle else ''),
                        'position': pos,
                        'aoi': ('https://openalex.org/'+self.inst_id in [i.get('id') for i in a['institutions']]),
                        'orcid': orcid,
                        'user': user
                    })

                pages = None
                if work['biblio']['first_page']:
                    pages = work['biblio']['first_page']
                    if work['biblio']['last_page'] and work['biblio']['last_page'] != pages:
                        pages += '-' + work['biblio']['last_page']

                # journal
                loc = work['primary_location']['source']
                # journal = loc['display_name']

                # date
                date = work['publication_date'].split('-')
                month = None
                day = None
                if len(date) >= 2:
                    month = int(date[1])
                if len(date) >= 3:
                    day = int(date[2])

                abstract = self.getAbstract(work.get('abstract_inverted_index'));
                element = {
                    'doi': doi,
                    'type': 'publication',
                    'subtype': typ,
                    'title': work['title'],
                    'year': work['publication_year'],
                    'abstract': abstract,
                    'month': month,
                    'day': day,
                    'authors': authors,
                    'pages': pages,
                    'openalex': work['id'].replace('https://openalex.org/', ''),
                    'pubmed': pubmed,
                    'open_access': work['open_access']['is_oa'],
                    'oa_status': work['open_access']['oa_status'],
                    'correction': False,
                    'epub': False
                }
                if (typ == 'others'):
                    element['doc_type'] = work['type'].title()

                if (typ == 'article'):
                    # print(element)
                    if not loc or not loc['issn']:
                        element['subtype'] = 'magazine'
                        element['magazine'] = loc.get('display_name') if loc else None
                    else:
                        journal = self.getJournal(loc['issn'])
                        element.update({
                            'volume': work['biblio']['volume'],
                            'issue': work['biblio']['issue'],
                            'journal': journal['journal'],
                            'journal_id': str(journal['_id'])
                        })
                        if (not element['volume']) and not element['issue']:
                            element['epub'] = True

                if (typ == 'chapter'):
                    element.update({
                        'book': loc['display_name'],
                        # 'edition': None,
                        # 'city': None,
                        # 'publisher': None,
                        # 'isbn': None
                    })

                for id, dupl in possible_dupl:
                    dist = ratio(dupl, element['title'])
                    # print(dist, dupl)
                    if (dist > 0.9):
                        element['duplicate'] = id
                        break
                yield element
                
    
    def queueJob(self):
        for element in self.get_works():
            self.osiris['queue'].insert_one(element)
    
    def importJob(self):
        for element in self.get_works():
            if element.get('duplicate'):
                continue
            self.osiris['activities'].insert_one(element)


if __name__ == '__main__':
    parser = OpenAlexParser()
    parser.queueJob()

File: php/OpenAlexParser.php
<?php

require_once 'DB.php';
require_once 'FullNameParser.php';
require_once 'Settings.php';

// use Diophila\OpenAlex;
// use HumanNameParser\Parser as HumanName;
// use Nlevenshtein\Levenshtein;

class OpenAlexParser
{
    public const TYPES = [
        "book-section" => "chapter",
        "monograph" => "book",
        "report-component" => "others",
        "report" => "others",
        "peer-review" => "others",
        "book-track" => "book",
        "journal-article" => "article",
        "article" => "article",
        "book-part" => "book",
        "other" => "others",
        "book" => "book",
        "journal-volume" => "article",
        "book-set" => "book",
        "reference-entry" => "others",
        "proceedings-article" => "others",
        "journal" => "others",
        "component" => "others",
        "book-chapter" => "chapter",
        "proceedings-series" => "others",
        "report-series" => "others",
        "proceedings" => "others",
        "database" => "others",
        "standard" => "others",
        "reference-book" => "book",
        "posted-content" => "others",
        "journal-issue" => "others",
        "dissertation" => "dissertation",
        "grant" => "others",
        "dataset" => "others",
        "book-series" => "book",
        "edited-book" => "book",
    ];
    private $inst_id;
    private $startyear;
    private $osiris;

    private $base_url = "https://api.openalex.org/";
    private $work_url = "works";
    private $source_url = "sources/";

    private $mail;
    private $NameParser;

    public function __construct($mail)
    {

        $this->mail = $mail;
        $DB = new DB();
        $this->osiris = $DB->db;

        $this->NameParser = new FullNameParser();

        $Settings = new Settings();
        $affiliation = $Settings->get('affiliation_details');

        $this->inst_id = $affiliation['openalex'] ?? null;
        if (!$this->inst_id) {
            throw new Exception("OpenAlex ID is missing.");
        }
        $this->startyear = $Settings->get('startyear');
    }

    private function getUserId($name, $orcid = null)
    {
        if ($orcid) {
            $user = $this->osiris->users->findOne(['orcid' => $orcid]);
            if ($user) {
                return $user->_id;
            }
        }
        $user = $this->osiris->users->findOne([
            'last' => $name['lname'],
            'first' => ['$regex' => '^' . $name['fname'] . '.*']
        ]);
        if ($user) {
            return $user->_id;
        }
        return null;
    }

    private function getAbstract($inverted_abstract)
    {
        if (!$inverted_abstract) return null;

        $abstract = [];
        foreach ($inverted_abstract as $word => $occurrence) {
            foreach ($occurrence as $oc) {
                $abstract[] = [$oc, $word];
            }
        }
        usort($abstract, function ($a, $b) {
            return $a[0] - $b[0];
        });
        return implode(" ", array_map(function ($i) {
            return $i[1];
        }, $abstract));
    }

    private function getJournal($issn)
    {
        $journal = $this->osiris->journals->findOne(['issn' => ['$in' => $issn]]);
        if ($journal) {
            return $journal;
        }

        $source = $this->get_single_source(end($issn), "issn");
        if (!$source) {
            return null;
        }
        if ($source['type'] != 'journal') {
            return [
                'magazine' => $source['display_name'],
                'publisher' => $source['host_organization_name'],
                'issn' => $source['issn'],
                'oa' => $source['is_oa'],
                // 'openalex' => str_replace('https://openalex.org/', '', $source['id'])
            ];
        }

        $new_journal = [
            'journal' => $source['display_name'],
            'abbr' => $source['abbreviated_title'],
            'publisher' => $source['host_organization_name'],
            'issn' => $source['issn'],
            'oa' => $source['is_oa'],
            'openalex' => str_replace('https://openalex.org/', '', $source['id'])
        ];
        $new_doc = $this->osiris->journals->insertOne($new_journal);
        $new_journal['_id'] = $new_doc->getInsertedId();
        return $new_journal;
    }

    public function parseWork($work)
    {
        // $result = ['msg' => '', 'status' => 'signal', 'link' => '', 'data' => []];

        if (!$work['doi'] || strpos($work['doi'], 'https://doi.org/') === false) {
            return ['title'=> $work['title'], 'msg' => 'DOI is missing or invalid.', 'status' => 'danger', 'link' => $work['id']];
        }

        if ($work['is_retracted']) {
            return ['title'=> $work['title'], 'msg' => 'Activity is retracted and was omitted.', 'status' => 'danger', 'link' => $work['doi']];
        }

        $pubmed = isset($work['ids']['pmid']) ? str_replace('https://pubmed.ncbi.nlm.nih.gov/', '', $work['ids']['pmid']) : null;
        $doi = str_replace('https://doi.org/', '', $work['doi']);

        if ($doi && $this->osiris->activities->countDocuments(['doi' => $doi]) > 0) {
            return ['title'=> $work['title'], 'msg' => "DOI exists and was omitted.", 'status' => 'danger', 'link' => $work['doi']];
        }
        if ($pubmed && $this->osiris->activities->countDocuments(['pubmed' => $pubmed]) > 0) {
            return ['title'=> $work['title'], 'msg' => "PubMed ID exists and was omitted.", 'status' => 'danger', 'link' => $work['ids']['pmid']];
        }

        // if ($this->osiris->queue->countDocuments(['doi' => $doi]) > 0) {
        //     return ['msg'=> "DOI is already in the queue.", 'status' => 'danger', 'link' => $work['doi']];
        // }
        if (!isset(OpenAlexParser::TYPES[$work['type']])) {
            if (isset(OpenAlexParser::TYPES[$work['type_crossref']])) {
                $work['type'] = $work['type_crossref'];
            } else {
                return ['title'=> $work['title'], 'msg' => "Activity type {$work['type']} is unknown.", 'status' => 'danger', 'link' => $work['doi']];
            }
        }

        $type = OpenAlexParser::TYPES[$work['type']];

        $authors = [];
        foreach ($work['authorships'] as $a) {
            $name = $this->NameParser->parse_name($a['author']['display_name']);
            $orcid = isset($a['author']['orcid']) ? str_replace('https://orcid.org/', '', $a['author']['orcid']) : null;
            $user = $this->getUserId($name, $orcid);
            $pos = $a['author_position'];
            if ($pos == 'middle' && isset($a['is_corresponding'])) {
                $pos = 'corresponding';
            }

            $approved = false;
            if ($user == $_SESSION['username']) {
                $approved = true;
            }
            $authors[] = [
                'last' => $name['lname'],
                'first' => $name['fname'] . ($name['initials'] ? ' ' . $name['initials'] : ''),
                'position' => $pos,
                'aoi' => in_array('https://openalex.org/' . $this->inst_id, array_column($a['institutions'], 'id')),
                'orcid' => $orcid,
                'user' => $user,
                'approved' => $approved
            ];
        }

        $pages = null;
        if (!empty($work['biblio']['first_page'])) {
            $pages = $work['biblio']['first_page'];
            if (!empty($work['biblio']['last_page']) && $work['biblio']['last_page'] != $pages) {
                $pages .= '-' . $work['biblio']['last_page'];
            }
        }

        $loc = $work['primary_location']['source'];

        $date = explode('-', $work['publication_date']);
        $month = isset($date[1]) ? intval($date[1]) : null;
        $day = isset($date[2]) ? intval($date[2]) : null;

        $abstract = $this->getAbstract($work['abstract_inverted_index']);
        $element = [
            'doi' => $doi,
            'type' => 'publication',
            'subtype' => $type,
            'title' => $work['title'],
            'year' => $work['publication_year'],
            'abstract' => $abstract,
            'month' => $month,
            'day' => $day,
            'authors' => $authors,
            'pages' => $pages,
            'openalex' => str_replace('https://openalex.org/', '', $work['id']),
            'pubmed' => $pubmed,
            'open_access' => $work['open_access']['is_oa'],
            'oa_status' => $work['open_access']['oa_status'],
            'correction' => false,
            'epub' => false
        ];
        if ($type == 'others') {
            $element['doc_type'] = ucfirst($work['type']);
        }

        if ($type == 'article') {
            if (!$loc || empty($loc['issn'])) {
                $element['subtype'] = 'magazine';
                $element['magazine'] = $loc ? $loc['display_name'] : null;
            } else {
                $journal = $this->getJournal($loc['issn']);
                if (isset($journal['magazine'])) {
                    $element['subtype'] = 'magazine';
                    $element = array_merge($element, $journal);
                } else {
                    $element = array_merge($element, [
                        'volume' => $work['biblio']['volume'],
                        'issue' => $work['biblio']['issue'],
                        'journal' => $journal['journal'],
                        'journal_id' => (string) $journal['_id']
                    ]);
                    if (empty($element['volume']) && empty($element['issue'])) {
                        $element['epub'] = true;
                    }
                }
            }
        }

        if ($type == 'chapter') {
            $element['book'] = $loc['display_name'];
        }

        $element['history'] = [
            ['date' => date('Y-m-d'), 'type' => 'imported', 'user' => $_SESSION['username']]
        ];

        $add = $this->osiris->activities->insertOne($element);
        $id = $add->getInsertedId();

        return [
            'msg' => 'Activity was added to OSIRIS.',
            'status' => 'success',
            'link' => $work['doi'],
            'title' => $work['title'],
            'id' => $id,
            'data' => $element
        ];
    }

    public function get_work($id, $idtype = 'doi', $ignoreDupl = true)
    {
        $work = $this->get_single_work($id, $idtype);
        return $this->parseWork($work);
    }

    public function get_works_from_author($id)
    {
        $filters = [
            'author.id:' . $id,
        ];

        $works = $this->get_list_of_works($filters);

        $i = 0;
        foreach ($works['results'] as $work) {
            $element = $this->parseWork($work);
            if ($element === false) continue;
            $i++;
            yield $element;
        }
    }


    private function callAPI($url)
    {
        $ch = curl_init($url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true); // important for redirects, e.g. issn search of source
        curl_setopt($ch, CURLOPT_HTTPHEADER, ['User-Agent: mailto:' . $this->mail]);
        $result = curl_exec($ch);
        curl_close($ch);
        return json_decode($result, true);
    }

    public function get_single_source($id, $idtype = null)
    {
        $filter = ($idtype !== null) ? $idtype . ':' . $id : $id;
        $url = $this->base_url . $this->source_url . $filter;
        return $this->callAPI($url);
    }
    public function get_single_work($id, $idtype = null)
    {
        $filter = ($idtype !== null) ? $idtype . ':' . $id : $id;
        $url = $this->base_url . $this->work_url . $filter;
        return $this->callAPI($url);
    }
    public function get_list_of_works($filters)
    {
        $url = $this->base_url . $this->work_url . '?filter=' . implode(',', $filters) . '&per-page=200';
        return $this->callAPI($url);
    }
}
//'https://api.openalex.org/works?filter=author.id:A5022180345'
